tools/query_breakdown/src/main/java/com/google/bigquery/LocationTracker.java

```
@@ -0,0 +1,92 @@
+/**
+ * This class tracks the original location of components in the query, thereby making sure that
+ * the error locations are correctly represented.
+ */
+public class LocationTracker {
+  /* we keep a double arraylist to represent the position of each character (line and column).
+     We can do this as the line number of the component will not change
+     (deletion and replacement won't change the line numbers)
+   */
+  private ArrayList<ArrayList<Integer>> location;
+
+  /**
+   * Constructor for the class
+   */
+  public LocationTracker() {
+    location = new ArrayList<>();
+  }
+
+  /**
+   * This method interacts with the InputReader and adds a pair to the location field that
+   * represents the position (x, y) in the original query. x and y are 1-indexed, so we
+   * adjust accordingly.
+   */
+  public void add(int x, int y) {
+    location.get(x - 1).add(y);
+  }
+
+  /**
+   * This method adds an empty line to the location field
+   */
+  public void addLine() {
+    location.add(new ArrayList<>());
+  }
+
+  /**
+   * This method removes an entry from location specified by (x, y) in the original query
+   */
+  public void remove(int x, int y) {
+    if (x > 0 && x <= location.size()) {
+      if (y > 0 && y <= location.get(x - 1).size()) {
+        location.get(x - 1).remove(y - 1);
+      }
+    }
+  }
+
+  /**
+   * This method gets the original position of the component in (x,y) of the intermediate query
+   */
+  public int getOriginalPosition(int x, int y) {
+    return location.get(x - 1).get(y - 1);
+  }
+
+  /**
+   * This method ensures that the location field is kept correctly despite the deletion.
+   */
+  public LocationTracker delete(int line, int startColumn, int endColumn) {
+    LocationTracker lt = cloneTracker();
+    for (int i = startColumn; i < endColumn + 1; i++) {
+     lt.remove(line, startColumn);
+    }
+    return lt;
+  }
+
+  /**
+   * To be implemented
+   */
+  public LocationTracker replace() {
+    return null;
```

throw UnsupportedOperationExceptions instead of return null.